keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,445 @@
+-->
+
+
+### PodsReady workload condition
+
+First, we introduce a new workload condition, called `PodsReady`, to indicate
+that the first batch of a workload's pods is scheduled, more precisely:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+
+The `PodsReady` condition is addded to the workload by the Kueue's Job
+reconciler in reaction to a status update of the corresponding Job. Note that,
+verifying if the condition should be added does not require an extra API call as
+the Kueue's Job reconciller already fetches the latest Job object at the
+beginning of the `Reconcile` function. Also note, there is no new condition
+introduced to the Job API.
+
+### Workload API
+
+We extend the Workload API with the new `startupRequirements` field to
+enable the new behavior.
+
+```golang
+// StartupRequirements describes workload startup requirements
+type StartupRequirements string
+
+const (
+  // QuotaAndProdsReady means that workload requires all pods to be running at
+  // the same time
+  StartupRequirementsQuotaAndProdsReady StartupRequirements = "QuotaAndProdsReady"
+  // QuotaOnly means that the only requirement for a workload to start is to
+  // get resources from the cluster quota
+  StartupRequirementsQuotaOnly StartupRequirements = "QuotaOnly"
+)
+
+// WorkloadSpec defines the desired state of Workload
+type WorkloadSpec struct {
+  ...
+  // StartupRequirements describes the workload requirement on the resource allocation.
+  // Possible values are:
+  // - QuotaAndProdsReady: means that workload requires allocation of resources
+  //   to support running of all pods at the same time,
+  // - QuotaOnly: means that the workload does not require all pods to be running
+  //   at the same time.
+  // Defaults to QuotaOnly.
+  StartupRequirements *string `json:"startupRequirements,omitempty"`
+}
+```
+
+### Job annotation
+
+We introduce a new Job annotation, called `kueue.x-k8s.io/workload-startup-requirements`,
+as a JobAPI counterpart of the `startupRequirements` Workload API field. When
+the annotation is present, its value is used as the field value during workload
+creation. Unknown value of the annotation results in leaving the field
+as null.
+
+### The mechanism
+
+When a workload with `startupRequirements=QuotaAndProdsReady` is nominated for admission,
+kueue scheduler blocks its admission (and admission of other workloads nominated
+in this cycle) until all already admitted workloads are in the `PodsReady`
+state. Kueue scheduler verifies this condition by a lookup to the cache
+of admitted workloads. When two or more such workloads are nomitated in a given
+cycle only one of them is admitted.
+
+Once such a workload is admitted, kueue's scheduler blocks admission of further
+workloads until the workload has the `PodsReady` condition. Similarly
+as in the previous case, Kueue's scheduler verifies the workload state by
+a lookup to the cache of admitted workloads.
```

Comment:
The fact that a workload is admitted is stored in the field: https://github.com/kubernetes-sigs/kueue/blob/aecac6d0168d9f204072b84eb48f5711425d1c24/apis/kueue/v1alpha2/workload_types.go#L41 so it can information in cache can be repopulated on the next startup of the kueue.